### ISSUE TITLE
AS generates invalid `.wast`

### ISSUE THREAD

**Issue by carllocos** [TOPLLab/WARDuino-VSCode#120]
I wrote an AssemnlyScript application (see `mirror-led.ts` and `config.ts` in zip) that cannot be compiled to `wasm`. In particular, `wat2wasm` fails to compile the obtained `wast` file to `wasm` due to several type errors. Some of the errors that I get are:`upload.wast:612:6: error: type mismatch in if true branch, expected [i32] but got []`, and `upload.wast:7658:4: error: type mismatch at end of function, expected [] but got [i32, i32]`. All the files generated by `asc` can be found in the directory output of the zip.

This compilation issue comes as a surprise because the `asc` generates alongside the `Wast` file also a `wasm` file indicating that no type error occurred in the process. So this may mean that our `wat2wasm` is behind some language features.

This type error issue only manifest when compiling via asc without any optimisation. Since we also want to debug without optimisation I will have to further investigate this issue.

[src_files.zip](https://github.com/TOPLLab/WARDuino-VSCode/files/11982141/src_files.zip)

**Comment by tolauwae** [comment]
It works on my machine with AssemblyScript `0.27.5`.

**Comment by carllocos** [comment]
@tolauwae did you also use no optimisation? (i.e., remove the `-O3s` from the `asc` command) because I also have version `0.27.5` and with optimisation the wasm gets generated but without optimisation, I get the type error.

**Comment by tolauwae** [comment]
I used the debug profile of the template which doesn't optimize. And it works. What is your exact asc command?

**Comment by carllocos** [comment]
full command
```
 asc path/to/src.ts --exportTable --disable bulk-memory --sourceMap --debug --outFile tmp/upload.wasm --textFile tmp/upload.wast`
```

**Something crucial I forgot to mention**
The type error occurring here above may occur unbeknownst to the user. For some reason, the plugin does not automatically stop even if the wat2wasm fails and thus proceeds with the next step i.e., flashing the `upload.wasm` to the MCU. Surprisingly this will succeed and the reason is that the plugin will search for a `tmp/upload.wasm` to flash  which will be present in the `tmp/` directory. This is because `asc` also generated besides the `wast` file a `upload.wasm` file that is put in the `tmp/` dir. In other words, from the user perspective flashing will be successful since the wasm of `asc` is used. The only way I noticed the `wat2wasm` failure is by putting a breakpoint on this [line](https://github.com/TOPLLab/WARDuino-VSCode/blob/f8a3cfd4831fd0046f4e7e5f04ecfb3284c0afce/src/CompilerBridges/WASMCompilerBridge.ts#L149). The `stderror` contained the type errors that I showed here above.

I also tried to use the template since it may not have the compilation error but I fail to properly configure it. Can you please confirm whether the following steps are correct:
- run `npm install` and `npm run build` in the root directory of the template
- add a `.vscode/launch.json` file into the root directory of the template
- change the `program` key of `launch.json` to point to the `main.ts` of the template
- replace this [line](https://github.com/TOPLLab/WARDuino-VSCode/blob/f8a3cfd4831fd0046f4e7e5f04ecfb3284c0afce/src/CompilerBridges/AssemblyScriptCompilerBridge.ts#L38) of the plugin with ```const command = `cd ${path.dirname(this.sourceFilePath.toString())} ; asc ${this.sourceFilePath} --config asconfig.json --target debug`;``` so that `asc` uses the config of the template

**Comment by tolauwae** [comment]
There is no need to recompile with `wat2wasm`. I haven't checked if there is a size difference, but I don't imagine the wasm from `asc` is larger than the wasm from `wat2wasm`.

**Comment by tolauwae** [comment]
Compiled file with your command also works for me.

**Comment by tolauwae** [comment]
> I also tried to use the template since it may not have the compilation error but I fail to properly configure it. Can you please confirm whether the following steps are correct:

Quickly tried it myself and AS doesn't seem to work right now in the plugin. Don't know what broke.

> replace this [line](https://github.com/TOPLLab/WARDuino-VSCode/blob/f8a3cfd4831fd0046f4e7e5f04ecfb3284c0afce/src/CompilerBridges/AssemblyScriptCompilerBridge.ts#L38) of the plugin with const command = `cd ${path.dirname(this.sourceFilePath.toString())} ; asc ${this.sourceFilePath} --config asconfig.json --target debug`; so that asc uses the config of the template

You don't have to change this. I tried it without changing any code on the main branch.

**Comment by tolauwae** [comment]
It wasn't working on my machine because the plugin used the globally installed `asc` instead of the local. #122 fixes this.

**Comment by tolauwae** [comment]
In other words, I still can't replicate your issue. The `wast2wasm` can compile the `.wast` generated by AS.

**Comment by tolauwae** [comment]
Nevermind. I must not have run it exactly the same. I get the error now as well.

**Comment by carllocos** [comment]
So I found the problem, the `asc` compiler was missing the `--runtime stub` option. 

To be honest I do not see immediately how that option relates to valid wast, especially because the invalid wast type was already previously reported and fixed by the maintainers of asc (see https://github.com/AssemblyScript/assemblyscript/issues/310).

I will push my fix as part of this PR https://github.com/TOPLLab/WARDuino-VSCode/pull/122. However, I now encountered another problem for which I still need to create a new issue (depending on the answer to this comment https://github.com/TOPLLab/WARDuino/pull/197#issuecomment-1633794707) and for which the fix maybe should be included also in the same PR.
